pkg/controllers/managementuser/resourcequota/resource_quota_sync.go

```diff
@@ -244,39 +237,45 @@ func (c *SyncController) getExistingLimitRange(ns *corev1.Namespace) (*corev1.Li
 	return limitRanger[0], nil
 }
 
-func (c *SyncController) getNamespaceResourceQuota(ns *corev1.Namespace) (*corev1.ResourceQuotaSpec, error) {
-	limit, err := getNamespaceResourceQuotaLimit(ns)
+// deriveRequestedResourceQuota tries to obtain the new namespace's resource quota limit and its quota spec.
+// It derives it by looking up the requested quota limit. If it's not found, then it looks up the project's default
+// quota for a namespace. If it's also not found, then the method returns an error.
+// If only the requested quota limit exists, then it is returned. If only the project's default namespace limit exists,
+// then it is returned. If both exist, then the two limits are merged, with requested limits having priority.
+func (c *SyncController) deriveRequestedResourceQuota(ns *corev1.Namespace) (*v32.ResourceQuotaLimit, *corev1.ResourceQuotaSpec, error) {
+	requested, err := getNamespaceResourceQuotaLimit(ns)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	if limit == nil {
-		limit = defaultResourceLimit
+
+	defaultQuota, err := getProjectNamespaceDefaultQuota(ns, c.ProjectLister)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	return convertResourceLimitResourceQuotaSpec(limit)
-}
+	var quotaLimit *v32.ResourceQuotaLimit
 
-func getDefaultQuotaSpec() (*corev1.ResourceQuotaSpec, error) {
-	return convertResourceLimitResourceQuotaSpec(defaultResourceLimit)
-}
+	if requested != nil && defaultQuota == nil {
+		quotaLimit = requested
+	} else if requested == nil && defaultQuota != nil {
+		quotaLimit = &defaultQuota.Limit
+	} else if requested != nil && defaultQuota != nil {
+		quotaLimit, err = completeQuota(requested, &defaultQuota.Limit)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		return nil, nil, errors.New("cannot parse resource quota limit: no limit is requested and no default exists")
+	}
 
-var defaultResourceLimit = &v32.ResourceQuotaLimit{
-	Pods:                   "0",
-	Services:               "0",
-	ReplicationControllers: "0",
-	Secrets:                "0",
-	ConfigMaps:             "0",
-	PersistentVolumeClaims: "0",
-	ServicesNodePorts:      "0",
-	ServicesLoadBalancers:  "0",
-	RequestsCPU:            "0",
-	RequestsMemory:         "0",
-	RequestsStorage:        "0",
-	LimitsCPU:              "0",
-	LimitsMemory:           "0",
+	newQuotaSpec, err := convertResourceLimitResourceQuotaSpec(requested)
```

Comment:
Wouldn't we want to convert the quotaLimit (which has gone through the parsing efforts above) rather than the requested value?

Reply:
Yes, this is my mistake! What's interesting is that none of my test scenarios produced incorrect behavior, which is odd...